### PR TITLE
PersonalizationApi Added.

### PR DESCRIPTION
**Pull request by akshays2112**
PersonalizationApi Added.

### CODE REVIEW

File: src/SpotifyApi.NetCore/PersonalizationApi.cs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        {
+        }
+        #endregion
+
+        public enum TimeRange
+        {
+            LongTerm,
+            MediumTerm,
+            ShortTerm
+        }
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/PersonalizationApi.cs]: Please move to Models folder and into its own file, i.e. Models/TimeRange.cs

File: src/SpotifyApi.NetCore/PersonalizationApi.cs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        /// <param name="limit">Optional. The maximum number of entities to return. Default: 20. Minimum: 1. Maximum: 50.</param>
+        /// <param name="offset">Optional. The index of the first object to return. Default: 0 (i.e., the first object). Use with limit to get the next set of objects.</param>
+        /// <returns>A Task that, once successfully completed, returns a full <see cref="PagedArtists"/> object.</returns>
+        /// <remarks>
+        /// https://developer.spotify.com/documentation/web-api/reference/personalization/get-users-top-artists-and-tracks/
+        /// </remarks>
+        public Task<PagedArtists> GetUsersTopArtists(
+            int limit = 20,
+            int offset = 0,
+            TimeRange timeRange = TimeRange.MediumTerm,
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/PersonalizationApi.cs:44]: `timeRange` and `accessToken` params need to be documented on all methods

File: src/SpotifyApi.NetCore/PersonalizationApi.cs [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+            if (limit < 1 || limit > 50) throw new
+                ArgumentException("A minimum of 1 and a maximum of 50 artist ids can be sent.");
+            if (offset < 0) throw new
+                ArgumentException("The offset must be an integer value greater than 0.");
+
+            var builder = new UriBuilder($"{BaseUrl}/me/top/artists");
+            builder.AppendToQuery("limit", limit);
+            builder.AppendToQuery("offset", offset);
+            builder.AppendToQuery("time_range", timeRange == TimeRange.ShortTerm ? "short_term" : 
+                (timeRange == TimeRange.LongTerm ? "long_term" : "medium_term"));
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/PersonalizationApi.cs]: This is a good idea. Please create an internal helper method to convert `TimeRange` to `string`, e.g.

```csharp
internal string TimeRangeString(TimeRange timeRange) => timeRange == TimeRange.ShortTerm ? "short_term" : 
                (timeRange == TimeRange.LongTerm ? "long_term" : "medium_term";
```

Also write Unit tests to assert that this mapping logic works as expected.

File: src/SpotifyApi.NetCore/PersonalizationApi.cs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        /// <summary>
+        /// Get the current user’s top tracks based on calculated affinity.
+        /// </summary>
+        /// <param name="limit">Optional. The maximum number of entities to return. Default: 20. Minimum: 1. Maximum: 50.</param>
+        /// <param name="offset">Optional. The index of the first object to return. Default: 0 (i.e., the first object). Use with limit to get the next set of objects.</param>
+        /// <returns>A Task that, once successfully completed, returns a full <see cref="PagedTracks"/> object.</returns>
+        /// <remarks>
+        /// https://developer.spotify.com/documentation/web-api/reference/personalization/get-users-top-artists-and-tracks/
+        /// </remarks>
+        public async Task<T> GetUsersTopTracks<T>(
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/PersonalizationApi.cs]: How about combining these two methods into one internal method with two public entry points? E.g.

```csharp
public Task<T> GetUsersTopTracks<T>(
    int limit = 20, 
    int offset = 0, 
    TimeRange timeRange = TimeRange.MediumTerm,
    string accessToken = null) => 
    GetUsersTopArtistsTracks(
        "tracks",
        limit: limit,
        offset: offset,
        timeRange: timeRange);

internal async Task<T> GetUsersTopArtistsTracks<T>(
    string type,
    int limit = 20, 
    int offset = 0, 
    TimeRange timeRange = TimeRange.MediumTerm,
    string accessToken = null)
    {
        // ...
    }
```

Sorry I did not think of this before. This would remove a fair bit of duplicate code 😃

File: src/SpotifyApi.NetCore.Tests/PersonalizationApiTests.cs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+            var api = new PersonalizationApi(http, accounts);
+
+            // act
+            PagedTracks response = await api.GetUsersTopTracks(
+                accessToken: testConfig.GetValue(typeof(string), "SpotifyUserBearerAccessToken").ToString());
+
+            // assert
+            Assert.IsNotNull(value: response);
+        }
+    }
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore.Tests/PersonalizationApiTests.cs:92]: Please add unit tests to assert TimeRange mapping logic works

File: src/SpotifyApi.NetCore/IPersonalizationApi.cs [hunk 6]
```diff
@@ -0,0 +1,77 @@
+﻿using SpotifyApi.NetCore.Models;
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/IPersonalizationApi.cs:1]: Please add missing param documentation

**Review by DanielLarsenNZ — COMMENTED** [review]
Really good! Just a couple of minor things 😃

**Comment by akshays2112** [comment]
Hey Daniel,

Will take care of it.

Thanks for your patience and time.

Best Regards,

Akshay Srinivasan

File: src/SpotifyApi.NetCore/SpotifyWebApi.cs [hunk 7]
```diff
@@ -172,8 +172,12 @@ private async Task<HttpResponseMessage> PostOrPut(string verb, Uri uri, object d
...
                 new AuthenticationHeaderValue("Bearer", accessToken ?? (await GetAccessToken()));
 
-            var content = data == null ? new StringContent("null") : new StringContent(JsonConvert.SerializeObject(data));
-            content.Headers.ContentType = new MediaTypeHeaderValue("application/json");
+            StringContent content = null;
+            if (data != null)
+            {
+                content = new StringContent(JsonConvert.SerializeObject(data));
+                content.Headers.ContentType = new MediaTypeHeaderValue("application/json");
+            }
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/SpotifyWebApi.cs]: As discussed offline, this change is too big for this PR (it impacts almost every API class in the library). Create a new PR for this and write tests to assert that none of the other calls will break.

File: src/SpotifyApi.NetCore/PersonalizationApi.cs [hunk 8]
```diff
@@ -92,34 +61,37 @@ public enum TimeRange
...
         /// <param name="limit">Optional. The maximum number of entities to return. Default: 20. Minimum: 1. Maximum: 50.</param>
         /// <param name="offset">Optional. The index of the first object to return. Default: 0 (i.e., the first object). Use with limit to get the next set of objects.</param>
+        /// <param name="timeRange">Optional. Over what time frame the affinities are computed. Valid values: long_term (calculated from several years of data and including all new data as it becomes available), medium_term (approximately last 6 months), short_term (approximately last 4 weeks). Default: medium_term.</param>
+        /// <param name="accessToken">The bearer token which is gotten during the authentication/authorization process.</param>
         /// <returns>A Task that, once successfully completed, returns a full <see cref="PagedTracks"/> object.</returns>
         /// <remarks>
         /// https://developer.spotify.com/documentation/web-api/reference/personalization/get-users-top-artists-and-tracks/
         /// </remarks>
-        public async Task<T> GetUsersTopTracks<T>(
+        public async Task<T> GetUsersTopArtistsOrTracks<T>(
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/PersonalizationApi.cs]: Make this `internal` and take it off the interface

File: src/SpotifyApi.NetCore/PersonalizationApi.cs [hunk 9]
```diff
@@ -92,34 +61,37 @@ public enum TimeRange
...
             if (offset < 0) throw new
                 ArgumentException("The offset must be an integer value greater than 0.");
 
-            var builder = new UriBuilder($"{BaseUrl}/me/top/tracks");
+            var builder = new UriBuilder($"{BaseUrl}/me/top/{type}s");
             builder.AppendToQuery("limit", limit);
             builder.AppendToQuery("offset", offset);
-            builder.AppendToQuery("time_range", timeRange == TimeRange.ShortTerm ? "short_term" :
-                (timeRange == TimeRange.LongTerm ? "long_term" : "medium_term"));
+            builder.AppendToQuery("time_range",  TimeRangeHelper.TimeRangeString(timeRange));
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/PersonalizationApi.cs:120]: Very nice! 😃

File: src/SpotifyApi.NetCore/FollowApi.cs [hunk 10]
```diff
@@ -31,69 +31,52 @@ public FollowApi(HttpClient httpClient) : base(httpClient)
...
+        /// <param name="type">Required. Spotify user or artist.</param>
+        /// <param name="userOrArtistIds">Required. A comma-separated list of the user Spotify IDs. A maximum of 50 user IDs can be sent in one request. A minimum of 1 user id is required.</param>
+        /// <param name="accessToken">The bearer token which is gotten during the authentication/authorization process.</param>
         /// <returns>bool[] an array of true or false values, in the same order in which the ids were specified.</returns>
         /// <remarks>
         /// https://developer.spotify.com/documentation/web-api/reference/follow/check-current-user-follows/
         /// </remarks>
-        public async Task<T> CheckCurrentUserFollowsUsers<T>(
-            string[] userIds,
+        public async Task<T> CheckCurrentUserFollowsArtistsOrUsers<T>(
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/FollowApi.cs]: Make this `internal` and take off the interface

File: src/SpotifyApi.NetCore/FollowApi.cs [hunk 11]
```diff
@@ -146,40 +131,48 @@ public FollowApi(HttpClient httpClient) : base(httpClient)
...
+        /// Add the current user as a follower of one or more Spotify users or artists.
+        /// </summary>
+        /// <param name="type">Required. Spotify user or artist.</param>
+        /// <param name="userIds">Required. A comma-separated list of the user Spotify IDs. A maximum of 50 user IDs can be sent in one request. A minimum of 1 user id is required.</param>
+        /// <param name="accessToken">The bearer token which is gotten during the authentication/authorization process.</param>
         /// <remarks>
         /// https://developer.spotify.com/documentation/web-api/reference/follow/follow-artists-users/
         /// </remarks>
-        public async Task FollowUsers(
+        public async Task FollowArtistsOrUsers(
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/FollowApi.cs]: Make `internal` and remove from interface

File: src/SpotifyApi.NetCore/FollowApi.cs [hunk 12]
```diff
@@ -255,41 +251,49 @@ public FollowApi(HttpClient httpClient) : base(httpClient)
...
+        /// <summary>
+        /// Remove the current user as a follower of one or more Spotify users or artists.
+        /// </summary>
+        /// <param name="type">Required. Spotify user or artist.</param>
+        /// <param name="userIds">A comma-separated list of the user Spotify IDs. A maximum of 50 user IDs can be sent in one request. A minimum of 1 user id is required. </param>
+        /// <param name="accessToken">The bearer token which is gotten during the authentication/authorization process.</param>
+        /// <remarks>
+        /// https://developer.spotify.com/documentation/web-api/reference/follow/unfollow-artists-users/
+        /// </remarks>
+        public async Task UnfollowArtistsOrUsers(
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/FollowApi.cs]: Make `internal` and remove from Interface

File: src/SpotifyApi.NetCore/IFollowApi.cs [hunk 13]
```diff
@@ -49,15 +38,18 @@ public interface IFollowApi
...
+        /// <param name="type">Required. Spotify user or artist.</param>
+        /// <param name="userOrArtistIds">Required. A comma-separated list of the user Spotify IDs. A maximum of 50 user IDs can be sent in one request. A minimum of 1 user id is required.</param>
+        /// <param name="accessToken">The bearer token which is gotten during the authentication/authorization process.</param>
         /// <returns>bool[] an array of true or false values, in the same order in which the ids were specified.</returns>
         /// <remarks>
         /// https://developer.spotify.com/documentation/web-api/reference/follow/check-current-user-follows/
         /// </remarks>
-        Task<T> CheckCurrentUserFollowsUsers<T>(
-            string[] userIds,
+        Task<T> CheckCurrentUserFollowsArtistsOrUsers<T>(
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/IFollowApi.cs]: Remove from Interface

File: src/SpotifyApi.NetCore/IFollowApi.cs [hunk 14]
```diff
@@ -99,25 +93,41 @@ public interface IFollowApi
...
+        /// <summary>
+        /// Add the current user as a follower of one or more Spotify users or artists.
+        /// </summary>
+        /// <param name="type">Required. Spotify user or artist.</param>
+        /// <param name="userIds">Required. A comma-separated list of the user Spotify IDs. A maximum of 50 user IDs can be sent in one request. A minimum of 1 user id is required.</param>
+        /// <param name="accessToken">The bearer token which is gotten during the authentication/authorization process.</param>
+        /// <remarks>
+        /// https://developer.spotify.com/documentation/web-api/reference/follow/follow-artists-users/
+        /// </remarks>
+        Task FollowArtistsOrUsers(
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/IFollowApi.cs]: Remove from Interface

File: src/SpotifyApi.NetCore/IFollowApi.cs [hunk 15]
```diff
@@ -185,20 +199,37 @@ public interface IFollowApi
...
+        /// <summary>
+        /// Remove the current user as a follower of one or more Spotify users or artists.
+        /// </summary>
+        /// <param name="type">Required. Spotify user or artist.</param>
+        /// <param name="userIds">A comma-separated list of the user Spotify IDs. A maximum of 50 user IDs can be sent in one request. A minimum of 1 user id is required. </param>
+        /// <param name="accessToken">The bearer token which is gotten during the authentication/authorization process.</param>
+        /// <remarks>
+        /// https://developer.spotify.com/documentation/web-api/reference/follow/unfollow-artists-users/
+        /// </remarks>
+        Task UnfollowArtistsOrUsers(
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/IFollowApi.cs]: Remove from Interface

File: src/SpotifyApi.NetCore/IFollowApi.cs [hunk 16]
```diff
@@ -22,23 +23,11 @@ public interface IFollowApi
...
-        /// <param name="artistIds">Required. A comma-separated list of the artist Spotify IDs. A maximum of 50 artist IDs can be sent in one request. A minimum of 1 artist id is required.</param>
-        /// <returns>bool[] an array of true or false values, in the same order in which the ids were specified.</returns>
-        /// <remarks>
-        /// https://developer.spotify.com/documentation/web-api/reference/follow/check-current-user-follows/
-        /// </remarks>
-        Task<T> CheckCurrentUserFollowsArtists<T>(
-            string[] artistIds,
-            string accessToken = null
-            );
-
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/IFollowApi.cs]: Don't remove from Interface

**Review by DanielLarsenNZ — CHANGES_REQUESTED** [review]
Sorry Akshay, I didn't mean for the helper methods to be public, please make them internal.

Please add tests for TimeRangeHelper, i.e. Helper/TimeRangeHelperTests.cs. These should be unit tests.

Please don't touch SpotifyWebApi in this PR. This is going to need a lot of testing. Let's tackle this together in a separate PR, as discussed.

File: src/SpotifyApi.NetCore.Tests/Helpers/TimeRangeHelperTests.cs [hunk 17]
```diff
@@ -0,0 +1,32 @@
+﻿using System;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using SpotifyApi.NetCore.Helpers;
+using SpotifyApi.NetCore.Models;
+
+namespace SpotifyApi.NetCore.Tests.Helpers
+{
+    [TestClass]
+    public class TimeRangeHelperTests
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore.Tests/Helpers/TimeRangeHelperTests.cs:9]: Awesome 👍

File: src/SpotifyApi.NetCore.Tests/EpisodeApiTests.cs [hunk 18]
```diff
@@ -10,44 +10,37 @@ namespace SpotifyApi.NetCore.Tests
     [TestClass]
     public class EpisodeApiTests
     {
-        [TestCategory("Integration")]
-        [TestMethod]
-        public async Task GetEpisode_EpisodeId_IsNotNull()
+        EpisodesApi api;
+        string bearerAccessToken;
+
+        public EpisodeApiTests()
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore.Tests/EpisodeApiTests.cs]: Reusing objects in test classes can have unintended consequences. If you want to keep your test code clean, put this code into a method called `Initialize()` with a `[TestInitialize]` attribute. This means the objects get re-initialized on every test, but you can keep your code clean.

File: src/SpotifyApi.NetCore.Tests/FollowApiTests.cs [hunk 19]
```diff
@@ -11,87 +11,64 @@ namespace SpotifyApi.NetCore.Tests
...
         {
-            // arrange
             var http = new HttpClient();
             IConfiguration testConfig = TestsHelper.GetLocalConfig();
+            bearerAccessToken = testConfig.GetValue(typeof(string), 
+                "SpotifyUserBearerAccessToken").ToString();
             var accounts = new AccountsService(http, testConfig);
+            api = new FollowApi(http, accounts);
+            usersApi = new UsersProfileApi(http, accounts);
+        }
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore.Tests/FollowApiTests.cs]: Reusing objects in test classes can have unintended consequences. If you want to keep your test code clean, put this code into a method called `Initialize()` with a `[TestInitialize]` attribute. This means the objects get re-initialized on every test, but you can keep your code clean.

File: src/SpotifyApi.NetCore.Tests/FollowApiTests.cs [hunk 20]
```diff
@@ -102,30 +79,23 @@ public async Task FollowArtists_Unfollowed_CheckUserFollowsTrue()
...
-            var accounts = new AccountsService(http, config);
-
-            var api = new FollowApi(http, accounts);
-
             // unfollow
-            await api.UnfollowArtists(artistIds, accessToken: accessToken);
+            await api.UnfollowArtists(artistIds, bearerAccessToken);
 
-            // checking if artists were unfollowed successfully
+            // checking if artists are already followed
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore.Tests/FollowApiTests.cs]: This comment is wrong, why did you change it? Here we are checking that the artist was unfollowed.

File: src/SpotifyApi.NetCore.Tests/FollowApiTests.cs [hunk 21]
```diff
@@ -102,30 +79,23 @@ public async Task FollowArtists_Unfollowed_CheckUserFollowsTrue()
...
             Assert.IsFalse((await api.CheckCurrentUserFollowsArtists(
                 artistIds,
-                accessToken: accessToken)).First(), 
-                "Artist should have been unfollowed at this point");
+                bearerAccessToken)).First(), 
+                "Artist is already followed.");
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore.Tests/FollowApiTests.cs]: This message is wrong, why did you change it? The artist should have been unfollowed at this point.

File: src/SpotifyApi.NetCore.Tests/FollowApiTests.cs [hunk 22]
```diff
@@ -102,30 +79,23 @@ public async Task FollowArtists_Unfollowed_CheckUserFollowsTrue()
...
 
             // act
-            await api.FollowArtists(artistIds, accessToken: accessToken);
+            await api.FollowArtists(artistIds, bearerAccessToken);
 
             // assert
-            // checking if artists were followed successfully
+            // checking if artists were unfollowed successfully
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore.Tests/FollowApiTests.cs]: Comment is wrong, why did you change it? Here we are checking to see if the artist was followed successfully.

File: src/SpotifyApi.NetCore.Tests/LibraryApiTests.cs [hunk 23]
```diff
@@ -11,263 +11,170 @@ namespace SpotifyApi.NetCore.Tests
...
+        public LibraryApiTests()
         {
-            // arrange
             var http = new HttpClient();
             IConfiguration testConfig = TestsHelper.GetLocalConfig();
+            bearerAccessToken = testConfig.GetValue(typeof(string),
+                "SpotifyUserBearerAccessToken").ToString();
             var accounts = new AccountsService(http, testConfig);
+            api = new LibraryApi(http, accounts);
+        }
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore.Tests/LibraryApiTests.cs]: Reusing objects in test classes can have unintended consequences. If you want to keep your test code clean, put this code into a method called `Initialize()` with a `[TestInitialize]` attribute. This means the objects get re-initialized on every test, but you can keep your code clean.

File: src/SpotifyApi.NetCore.Tests/PersonalizationApiTests.cs [hunk 24]
```diff
@@ -0,0 +1,87 @@
...
+
+        public PersonalizationApiTests()
+        {
+            var http = new HttpClient();
+            IConfiguration testConfig = TestsHelper.GetLocalConfig();
+            bearerAccessToken = testConfig.GetValue(typeof(string),
+                "SpotifyUserBearerAccessToken").ToString();
+            var accounts = new AccountsService(http, testConfig);
+            api = new PersonalizationApi(http, accounts);
+        }
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore.Tests/PersonalizationApiTests.cs]: Reusing objects in test classes can have unintended consequences. If you want to keep your test code clean, put this code into a method called `Initialize()` with a `[TestInitialize]` attribute. This means the objects get re-initialized on every test, but you can keep your code clean.

File: src/SpotifyApi.NetCore.Tests/PersonalizationApiTests.cs [hunk 25]
```diff
@@ -0,0 +1,87 @@
...
+
+        [TestCategory("Integration")]
+        [TestMethod]
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore.Tests/PersonalizationApiTests.cs:29]: Add `[TestCategory("User")]` here because this test requires User auth.

File: src/SpotifyApi.NetCore.Tests/PersonalizationApiTests.cs [hunk 26]
```diff
@@ -0,0 +1,87 @@
...
+        public async Task GetUserTopArtists_TimeRange_LongTerm_IsNotNull()
+        {
+            // assert
+            Assert.IsNotNull(value: await api.GetUsersTopArtists(
+                timeRange: TimeRange.LongTerm,
+                accessToken: bearerAccessToken));
+        }
+
+        [TestCategory("Integration")]
+        [TestMethod]
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore.Tests/PersonalizationApiTests.cs:40]: Add `[TestCategory("User")]` here because this test requires User auth.

File: src/SpotifyApi.NetCore.Tests/PersonalizationApiTests.cs [hunk 27]
```diff
@@ -0,0 +1,87 @@
...
+        public async Task GetUserTopArtists_TimeRange_MediumTerm_IsNotNull()
+        {
+            // assert
+            Assert.IsNotNull(value: await api.GetUsersTopArtists(
+                timeRange: TimeRange.MediumTerm,
+                accessToken: bearerAccessToken));
+        }
+
+        [TestCategory("Integration")]
+        [TestMethod]
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore.Tests/PersonalizationApiTests.cs:51]: Add `[TestCategory("User")]` here because this test requires User auth.

File: src/SpotifyApi.NetCore.Tests/PersonalizationApiTests.cs [hunk 28]
```diff
@@ -0,0 +1,87 @@
...
+        public async Task GetUserTopArtists_TimeRange_ShortTerm_IsNotNull()
+        {
+            // assert
+            Assert.IsNotNull(value: await api.GetUsersTopArtists(
+                timeRange: TimeRange.ShortTerm,
+                accessToken: bearerAccessToken));
+        }
+
+        [TestCategory("Integration")]
+        [TestMethod]
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore.Tests/PersonalizationApiTests.cs:62]: Add `[TestCategory("User")]` here because this test requires User auth.

File: src/SpotifyApi.NetCore.Tests/PersonalizationApiTests.cs [hunk 29]
```diff
@@ -0,0 +1,87 @@
...
+        public async Task GetUserTopTracks_TimeRange_LongTerm_IsNotNull()
+        {
+            // assert
+            Assert.IsNotNull(value: await api.GetUsersTopTracks(
+                timeRange: TimeRange.LongTerm,
+                accessToken: bearerAccessToken));
+        }
+
+        [TestCategory("Integration")]
+        [TestMethod]
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore.Tests/PersonalizationApiTests.cs:73]: Add `[TestCategory("User")]` here because this test requires User auth.

File: src/SpotifyApi.NetCore.Tests/PersonalizationApiTests.cs [hunk 30]
```diff
@@ -0,0 +1,87 @@
...
+        public async Task GetUserTopTracks_TimeRange_MediumTerm_IsNotNull()
+        {
+            // assert
+            Assert.IsNotNull(value: await api.GetUsersTopTracks(
+                timeRange: TimeRange.MediumTerm,
+                accessToken: bearerAccessToken));
+        }
+
+        [TestCategory("Integration")]
+        [TestMethod]
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore.Tests/PersonalizationApiTests.cs:84]: Add `[TestCategory("User")]` here because this test requires User auth.

File: src/SpotifyApi.NetCore/LibraryApi.cs [hunk 31]
```diff
@@ -25,121 +25,91 @@ public LibraryApi(HttpClient httpClient) : base(httpClient)
...
+        /// <param name="accessToken">The bearer token which is gotten during the authentication/authorization process.</param>
         /// <returns>bool[] an array of true or false values, in the same order in which the ids were specified.</returns>
         /// <remarks>
+        /// https://developer.spotify.com/documentation/web-api/reference/library/check-users-saved-albums/
+        /// https://developer.spotify.com/documentation/web-api/reference/library/check-users-saved-shows/
         /// https://developer.spotify.com/documentation/web-api/reference/library/check-users-saved-tracks/
         /// </remarks>
-        public async Task<T> CheckUsersSavedTracks<T>(
-            string[] trackIds,
+        public async Task<T> CheckUsersSavedAlbumsOrShowsOrTracks<T>(
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/LibraryApi.cs]: This should be `internal`

File: src/SpotifyApi.NetCore/LibraryApi.cs [hunk 32]
```diff
@@ -149,31 +119,58 @@ public LibraryApi(HttpClient httpClient) : base(httpClient)
...
         /// <param name="market">Optional. An ISO 3166-1 alpha-2 country code or the string from_token or the string from_token <see cref="SpotifyCountryCodes" />. Provide this parameter if you want to apply Track Relinking.</param>
-        /// <returns>A Task that, once successfully completed, returns a full <see cref="PagedAlbums"/> object.</returns>
+        /// <param name="accessToken">The bearer token which is gotten during the authentication/authorization process.</param>
+        /// <returns>A Task that, once successfully completed, returns a full <see cref="PagedAlbums"/> or <see cref="PagedTracks"/> object depending on the param type value.</returns>
         /// <remarks>
         /// https://developer.spotify.com/documentation/web-api/reference/library/get-users-saved-albums/
+        /// https://developer.spotify.com/documentation/web-api/reference/library/get-users-saved-tracks/
         /// </remarks>
-        public async Task<T> GetCurrentUsersSavedAlbums<T>(
+        public async Task<T> GetCurrentUsersSavedAlbumsOrTracks<T>(
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/LibraryApi.cs]: This should be `internal`

File: src/SpotifyApi.NetCore/ILibraryApi.cs [hunk 33]
```diff
@@ -77,28 +55,36 @@ public interface ILibraryApi
...
         /// <remarks>
+        /// https://developer.spotify.com/documentation/web-api/reference/library/check-users-saved-albums/
+        /// https://developer.spotify.com/documentation/web-api/reference/library/check-users-saved-shows/
         /// https://developer.spotify.com/documentation/web-api/reference/library/check-users-saved-tracks/
         /// </remarks>
-        Task<T> CheckUsersSavedTracks<T>(
-            string[] trackIds,
+        Task<T> CheckUsersSavedAlbumsOrShowsOrTracks<T>(
+            string type,
+            string[] albumOrShowOrTrackIds,
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/ILibraryApi.cs]: Don't put `CheckUsersSavedAlbumsOrShowsOrTracks` on the interface. This is should be `internal`. Leave the generic methods on the Interface

File: src/SpotifyApi.NetCore/ILibraryApi.cs [hunk 34]
```diff
@@ -109,103 +95,110 @@ public interface ILibraryApi
...
+        /// https://developer.spotify.com/documentation/web-api/reference/library/get-users-saved-tracks/
         /// </remarks>
-        Task<PagedShows> GetUsersSavedShows(
+        Task<T> GetCurrentUsersSavedAlbumsOrTracks<T>(
+            string type,
             int limit = 20,
             int offset = 0,
+            string market = null,
             string accessToken = null
             );
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/ILibraryApi.cs]: `GetCurrentUsersSavedAlbumsOrTracks` should be an internal method and should not be on the Interface.

File: src/SpotifyApi.NetCore/FollowApi.cs [hunk 35]
```diff
@@ -204,7 +219,7 @@ public FollowApi(HttpClient httpClient) : base(httpClient)
                     ArgumentNullException("playlistId");
 
             var builder = new UriBuilder($"{BaseUrl}/playlists/{playlistId}/followers");
-            await Put(builder.Uri, isPublic, accessToken);
+            await Put(builder.Uri, new { @public = isPublic }, accessToken);
```
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/FollowApi.cs]: This looks like a breaking change to a Released API - have you got a test that asserts that this still works?

**Review by DanielLarsenNZ — CHANGES_REQUESTED** [review]

File: src/SpotifyApi.NetCore/FollowApi.cs (hunk 35 above)
Comment (akshays2112) [reply src/SpotifyApi.NetCore/FollowApi.cs]: Hi Daniel,

Thank you for taking time out for this.  Will triple check.

Thanks for your patience and time.

Best Regards,

Akshay Srinivasan

File: src/SpotifyApi.NetCore.Tests/EpisodeApiTests.cs (hunk 18 above)
Comment (akshays2112) [reply src/SpotifyApi.NetCore.Tests/EpisodeApiTests.cs]: Hi Daniel,

Thanks did not know this.  Learning a lot as I knew I would from you. Working on all these changes as fast as I can.

Thanks for your patience and time.

Best Regards,

Akshay Srinivasan

File: src/SpotifyApi.NetCore.Tests/FollowApiTests.cs (hunk 20 above)
Comment (akshays2112) [reply src/SpotifyApi.NetCore.Tests/FollowApiTests.cs]: Hi Daniel,

To be honest I was drinking a bit of rum and coke while doing all the changes.  So I screwed this one up.

Thanks for your patience and time.

Best Regards,

Akshay Srinivasan

File: src/SpotifyApi.NetCore/FollowApi.cs (hunk 35 above)
Comment (akshays2112) [reply src/SpotifyApi.NetCore/FollowApi.cs]: Hi Daniel,

My logic was that the SpotifyAPI documentation expects the value's property name to be the Json string "public".  While somehow it worked it is not perfectly correct the original code. This is why I changed it.  Working on unit tests to verify this works perfectly well.

Thanks for your patience and time.

Best Regards,

Akshay Srinivasan

Comment (akshays2112) [reply src/SpotifyApi.NetCore/FollowApi.cs]: Hi Daniel,

One scenario of why this would have worked originally is that the parameter value on Spotify's end is accessing via index and hence wrong parameter name as per the documentation works anyways on this API call.

Thanks for your patience and time.

Best Regards,

Akshay Srinivasan

Comment (akshays2112) [reply src/SpotifyApi.NetCore/FollowApi.cs]: Hi Daniel,

My suspicion was right.  The parameter name can be anything and does not have to be called "public" as per the SpotifyAPI documentation as on their end they are not using the parameter name to get to the value.  Technically this code change is not required at all.

How do I know this I changed @public = ... to @xyz = ... and it passed with no errors.

So, for the time being, I will revert this change as this works always until they change the code in their API to not use the index but use the parameter name as per their documentation to access the value.

So as of this message I am done from my side. The ball is back in your court Daniel ♥♥♥.

Thanks for your patience and time.

Best Regards,

Akshay Srinivasan

File: src/SpotifyApi.NetCore.Tests/FollowApiTests.cs (hunk 20 above)
Comment (DanielLarsenNZ) [reply src/SpotifyApi.NetCore.Tests/FollowApiTests.cs]: Ha ha ha fair enough! 😄

**Review by DanielLarsenNZ — APPROVED** [review]

**Review by DanielLarsenNZ — APPROVED** [review]
